packages/cojson-storage-indexeddb/src/idbClient.ts

```diff
@@ -0,0 +1,234 @@
+import { CojsonInternalTypes, SessionID } from "cojson";
+import { SyncPromise } from "./syncPromises";
+import RawCoID = CojsonInternalTypes.RawCoID;
+import Transaction = CojsonInternalTypes.Transaction;
+import Signature = CojsonInternalTypes.Signature;
+
+export type CoValueRow = {
+  id: CojsonInternalTypes.RawCoID;
+  header: CojsonInternalTypes.CoValueHeader;
+};
+
+export type StoredCoValueRow = CoValueRow & { rowID: number };
+
+export type TransactionRow = {
+  ses: number;
+  idx: number;
+  tx: CojsonInternalTypes.Transaction;
+};
+
+export type SignatureAfterRow = {
+  ses: number;
+  idx: number;
+  signature: CojsonInternalTypes.Signature;
+};
+
+export type SessionRow = {
+  coValue: number;
+  sessionID: SessionID;
+  lastIdx: number;
+  lastSignature: CojsonInternalTypes.Signature;
+  bytesSinceLastSignature?: number;
+};
+
+export type StoredSessionRow = SessionRow & { rowID: number };
+
+export class IDBClient {
+  private db;
+
+  currentTx:
+    | {
+        id: number;
+        tx: IDBTransaction;
+        stores: {
+          coValues: IDBObjectStore;
+          sessions: IDBObjectStore;
+          transactions: IDBObjectStore;
+          signatureAfter: IDBObjectStore;
+        };
+        startedAt: number;
+        pendingRequests: ((txEntry: {
+          stores: {
+            coValues: IDBObjectStore;
+            sessions: IDBObjectStore;
+            transactions: IDBObjectStore;
+            signatureAfter: IDBObjectStore;
+          };
+        }) => void)[];
+      }
+    | undefined;
+
+  currentTxID = 0;
+
+  constructor(db: IDBDatabase) {
+    this.db = db;
+  }
+
+  makeRequest<T>(
+    handler: (stores: {
+      coValues: IDBObjectStore;
+      sessions: IDBObjectStore;
+      transactions: IDBObjectStore;
+      signatureAfter: IDBObjectStore;
+    }) => IDBRequest,
+  ): SyncPromise<T> {
+    return new SyncPromise((resolve, reject) => {
+      let txEntry = this.currentTx;
+
+      const requestEntry = ({
+        stores,
+      }: {
+        stores: {
+          coValues: IDBObjectStore;
+          sessions: IDBObjectStore;
+          transactions: IDBObjectStore;
+          signatureAfter: IDBObjectStore;
+        };
+      }) => {
+        const request = handler(stores);
+        request.onerror = () => {
+          console.error("Error in request", request.error);
+          this.currentTx = undefined;
+          reject(request.error);
+          // TODO: recover pending requests in new tx
+        };
+        request.onsuccess = () => {
+          const value = request.result as T;
+          resolve(value);
+
+          const next = txEntry!.pendingRequests.shift();
+
+          if (next) {
+            next({ stores });
+          } else {
+            if (this.currentTx === txEntry) {
+              this.currentTx = undefined;
+            }
+          }
+        };
+      };
+
+      if (!txEntry || performance.now() - txEntry.startedAt > 20) {
```

Comment:
Honestly I didn't change anything in that function as it's out of our focus. Just left it as it is as it works and IDB is not our priority. Wdyt?

Reply:
What this implements is "transaction batching" which generally helps with IndexedDB performance. I'm not sure if we're doing it the best way but it would be good to at least comment that this is what we're doing.

Reply:
Added comment